File: apps/api/app/schemas/interview.py
import uuid
from datetime import datetime

from pydantic import BaseModel, ConfigDict, Field

from app.models.interview import InterviewMode, InterviewStatus


class InterviewCreateRequest(BaseModel):
    mode: InterviewMode
    target_role: str | None = Field(default=None, max_length=255)


class InterviewAnswerRequest(BaseModel):
    question_id: uuid.UUID
    answer_text: str = Field(min_length=1, max_length=10_000)
    response_time_seconds: int = Field(ge=0, le=3_600)


class InterviewEvaluationRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    correctness_score: float
    depth_score: float
    communication_score: float
    feedback: str


class InterviewAnswerRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    answer_text: str
    response_time_seconds: int
    created_at: datetime
    evaluation: InterviewEvaluationRead | None = None


class InterviewQuestionRead(BaseModel):
    """One question in a session, with its answer/evaluation if already submitted. The
    frontend/route derive "the current question" as the first item with `answer is None` —
    there's no separate `current_question` field, matching the "one response has everything"
    pattern RAG/roadmap already use rather than a second, potentially-inconsistent field."""

    id: uuid.UUID
    question_text: str
    category: str
    order_index: int
    answer: InterviewAnswerRead | None = None


class InterviewRead(BaseModel):
    """List-view projection — `GET /api/v1/interviews` (history)."""

    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    mode: InterviewMode
    target_role: str | None
    status: InterviewStatus
    overall_score: float | None
    created_at: datetime


class InterviewDetailRead(InterviewRead):
    """`POST /api/v1/interviews`, `GET /api/v1/interviews/{id}`,
    `POST /api/v1/interviews/{id}/answer` — every route that returns a full session. `questions`
    isn't an ORM relationship — populated by the route from a fresh, ordered query joining
    answers/evaluations (app/services/interviews.py), same reasoning as `LearningRoadmapRead`."""

    questions: list[InterviewQuestionRead] = Field(default_factory=list)


class InterviewAnalyticsRecent(BaseModel):
    interview_id: uuid.UUID
    mode: InterviewMode
    overall_score: float
    completed_at: datetime


class InterviewAnalyticsRead(BaseModel):
    """`GET /api/v1/interviews/analytics` — real SQL aggregates over the current user's own
    completed sessions only. All fields `None`/empty when the user has no completed interviews
    yet, never a fabricated zero that could be misread as "you scored 0"."""

    total_completed: int
    average_overall_score: float | None
    average_correctness_score: float | None
    average_depth_score: float | None
    average_communication_score: float | None
    recent: list[InterviewAnalyticsRecent] = Field(default_factory=list)

File: apps/web/lib/types/interview.ts
export type InterviewMode =
  | "technical"
  | "behavioral"
  | "hr"
  | "system_design"
  | "ml"
  | "data_science";

export type InterviewStatus = "in_progress" | "completed" | "abandoned";

export const INTERVIEW_MODE_LABELS: Record<InterviewMode, string> = {
  technical: "Technical",
  behavioral: "Behavioral",
  hr: "HR",
  system_design: "System Design",
  ml: "Machine Learning",
  data_science: "Data Science",
};

export type InterviewEvaluation = {
  correctness_score: number;
  depth_score: number;
  communication_score: number;
  feedback: string;
};

export type InterviewAnswer = {
  answer_text: string;
  response_time_seconds: number;
  created_at: string;
  evaluation: InterviewEvaluation | null;
};

export type InterviewQuestion = {
  id: string;
  question_text: string;
  category: string;
  order_index: number;
  answer: InterviewAnswer | null;
};

/** List-view projection — `GET /api/v1/interviews` (history). */
export type Interview = {
  id: string;
  mode: InterviewMode;
  target_role: string | null;
  status: InterviewStatus;
  overall_score: number | null;
  created_at: string;
};

/** `POST /api/v1/interviews`, `GET /api/v1/interviews/{id}`, `POST .../answer` — every route
 * that returns a full session, including every question asked so far with its answer/evaluation
 * if already submitted. The "current question" is derived client-side as the first item with
 * `answer === null` (docs/API.md §5, Phase 11). */
export type InterviewDetail = Interview & {
  questions: InterviewQuestion[];
};

export type InterviewAnalyticsRecent = {
  interview_id: string;
  mode: InterviewMode;
  overall_score: number;
  completed_at: string;
};

/** `GET /api/v1/interviews/analytics` — real aggregates over the user's own completed sessions
 * only. All averages `null`/`recent` empty when the user has no completed interviews yet. */
export type InterviewAnalytics = {
  total_completed: number;
  average_overall_score: number | null;
  average_correctness_score: number | null;
  average_depth_score: number | null;
  average_communication_score: number | null;
  recent: InterviewAnalyticsRecent[];
};
